Snabble/UI/Payment/ExitTokenPoller.swift

```diff
@@ -0,0 +1,77 @@
+//
+//  ExitTokenPoller.swift
+//
+//  Copyright © 2021 snabble. All rights reserved.
+//
+
+final class ExitTokenPoller {
+    static let shared = ExitTokenPoller()
+
+    private var processTimer: Timer?
+    private var sessionTask: URLSessionTask?
+    private var project: Project?
+    private var process: CheckoutProcess?
+
+    private var completion: ((CheckoutProcess, [String: Any]?) -> Void)?
+
+    private init() {}
+
+    func startPolling(_ project: Project, _ process: CheckoutProcess, completion: @escaping (CheckoutProcess, [String: Any]?) -> Void) {
+        if self.process == nil && self.process == nil {
```

Comment:
Falls die Bedingung nicht erfüllt ist, ist der Entwickler in einem Deadlock, da `completion()` nicht aufgerufen wird.

Reply:
Ja, siehe oben, aber das hier ist ein Schutz dagegen dass nicht zwei dieser Poller gleichzeitig laufen.

Reply:
Was hältst du davon die `completion()` zu einem `Swift.Result` zu ändern um mit einem Failure zu antworten? So wäre der Entwickler nicht in einem potentiellen Deadlock.

Reply:
Ich verstehe nicht wo du einen Deadlock siehst. Es gibt hier weit und breit keine Locks, und dieser Test ist wie gesagt dazu da, dass man nicht irrtümlich zwei dieser Poller gleichzeitig startet, indem nämlich sämtliche Start-Requests nach dem 1. einfach ignoriert werden. Wenn das passieren sollte ist der eigentliche Bug ganz woanders, aber die completion() wird ja durch den ersten Lauf irgendwann aufgerufen.

Reply:
Der Deadlock entsteht beim doppelten Aufruf:

```swift
let project: Project = ...
let process: CheckoutProcess = ...

ExitTokenPoller.shared.startPolling(project, process, completion: { _ in 
    // Wird irgendwann aufgerufen
}

ExitTokenPoller.shared.startPolling(project, process, completion: { _ in 
    // Wird niemals aufgerufen
}
```

Falls der Entwickler nun auf die Response durch den completionHandler wartet wird er enttäuscht und niemals einen Aufruf erhalten. Ist es nicht verwirrend das `completion` nicht aufgerufen wird?

Idee wäre beispielsweise ein Result mit einem Error _"Already polling for project with process"_.
```swift
enum Error: Swift.Error {
  case .alreadyPolling(Project, Process)
}
```

Reply:
OK das ist nicht meine Definition/Interpretation von Deadlock, aber egal. 

Solcher Code wäre ein Bug und darf so nie passieren/geschrieben werden. Anstelle des simplen `return` könnte in der Funktion genausogut eine assertion oder ein `fatalError` passieren, und da seh ich nicht wieso wir ein Errorhandling für so eine this-cant-happen-Situation einbauen sollten.

Reply:
`fatalError` klingt auch in Ordnung 👍